c++/src/kj/compat/http.c++

```diff
@@ -5709,17 +5709,36 @@ public:
       }).attach(kj::mv(address));
     }).split();
 
+    auto connection = kj::refcountedWrapper(kj::newPromisedStream(kj::mv(kj::get<1>(split))));
+
+    #if KJ_HAS_OPENSSL
+    // XXX: I'm uncertain whether this is the best approach.
+    kj::Own<kj::AsyncIoStream> ref1 = connection->addWrappedRef();
+    KJ_IF_MAYBE(tlsNet, tlsNetwork) {
+      kj::TlsContext& tlsContext = tlsNet->getTlsContext();
+      KJ_IF_MAYBE(tlsStarter, settings.tlsStarter) {
+        Function<kj::Own<kj::AsyncIoStream>(kj::StringPtr)> cb =
+            [&tlsContext, ref1 = kj::mv(ref1)](kj::StringPtr expectedServerHostname) mutable {
+          kj::Promise<kj::Own<kj::AsyncIoStream>> secureStream =
+              tlsContext.wrapClient(kj::mv(ref1), expectedServerHostname);
+          return kj::newPromisedStream(kj::mv(secureStream));
+        };
+        *tlsStarter = kj::mv(cb);
```

Comment:
It looks like the type of `settings.tlsStarter` is `kj::Maybe<TlsStarterCallback&>`... I think this assignment will just capture a reference to the `Function` object in this scope, such that `cb` 's contents will be destroyed at the end of the scope, making the reference invalid.  Probably want the type of `settings.tlsStarter` to be `kj::Maybe<TlsStarterCallback>` instead?

Also, looks like the type of `TlsStarterCallback` itself contains a Maybe, which is possibly redundant?

Reply:
The idea here is that the caller of `connect` specifies a reference to a `Function` so that `connect` can fill that with a callback. As far as I understand I am moving the `Function` into that reference here. From my testing it seems to be working but maybe I just got lucky and am missing something.

The reason `TlsStarterCallback` contains a Maybe is because the `connect` might not fill the callback (if for example there is no TLS support). This is `settings.tlsStarter` is two Maybes, one because the caller might not want the callback to be filled anywhere (in which case they set this option to `nullptr`) and the second because of what I outlined previously.

Indeed, the API is ugly. We had a discussion about how to handle this and this was what we landed on. If anyone has better ideas please feel free to share.

Good point on the assert.

Reply:
A differently-ugly option could be to make `ConnectRequest::connection` a `kj::OneOf<kj::Own<kj::AsyncIoStream>, kj::Own<kj::TlsStartableAsyncIoStream>>`, the latter type growing a `startTls()` member function which is callable once and either upgrades the connection in situ, or returns a new AsyncIoStream, invalidating the old one. That could potentially make lifetimes clearer.

It'd be possible to avoid the OneOf with a separate `connectWithStartTls()` function (perhaps only available if KJ_USE_OPENSSL is true) which returns the one-and-only type.